File: frontend/tests/contract/api/auth.contract.test.ts
// Authentication API Contract Tests
// Validates API integration against contracts/openapi.yaml

import { describe, it, expect, beforeAll, afterAll, afterEach } from 'vitest';
import { setupServer } from 'msw/node';
import { http, HttpResponse } from 'msw';

// API client (not yet implemented - tests will fail)
import { apiClient } from '@/services/api';
import type { LoginResponse, UserDTO, ErrorResponse } from '@/types/api';

// =============================================================================
// MSW Mock Server Setup
// =============================================================================

const server = setupServer();

beforeAll(() => server.listen({ onUnhandledRequest: 'error' }));
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

// =============================================================================
// POST /api/auth/login - LDAP Authentication
// =============================================================================

describe('POST /api/auth/login - LDAP', () => {
  it('returns LoginResponse schema on successful LDAP authentication', async () => {
    // Arrange: Mock successful LDAP authentication
    server.use(
      http.post('http://localhost:7075/api/auth/login', async ({ request }) => {
        const body = await request.json() as any;

        if (body.username === 'dechawat' && body.password === 'TestPassword123') {
          return HttpResponse.json<LoginResponse>({
            token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyaWQiOjQyLCJ1c2VybmFtZSI6ImRlY2hhd2F0In0.abc123',
            user: {
              userid: 42,
              username: 'dechawat',
              firstName: 'Dechawat',
              lastName: 'Wongsirasawat',
              department: 'Warehouse',
              authSource: 'LDAP',
              permissions: ['putaway', 'picking', 'partial-picking']
            }
          });
        }

        return HttpResponse.json(
          { error: { code: 'AUTH_INVALID_CREDENTIALS', message: 'Invalid credentials', correlationId: 'test' } },
          { status: 401 }
        );
      })
    );

    // Act
    const response = await apiClient.login('dechawat', 'TestPassword123');

    // Assert: Validate LoginResponse schema
    expect(response).toBeDefined();
    expect(response.token).toMatch(/^eyJ/); // JWT starts with eyJ
    expect(response.user).toBeDefined();
    expect(response.user.userid).toBe(42);
    expect(response.user.username).toBe('dechawat');
    expect(response.user.authSource).toBe('LDAP');
    expect(response.user.permissions).toContain('partial-picking');
  });

  it('returns ErrorResponse schema on invalid LDAP credentials', async () => {
    // Arrange: Mock failed authentication
    server.use(
      http.post('http://localhost:7075/api/auth/login', () => {
        return HttpResponse.json<ErrorResponse>(
          {
            error: {
              code: 'AUTH_INVALID_CREDENTIALS',
              message: 'Invalid credentials',
              correlationId: 'test-correlation-id',
              details: null
            }
          },
          { status: 401 }
        );
      })
    );

    // Act & Assert
    await expect(
      apiClient.login('dechawat', 'WrongPassword')
    ).rejects.toMatchObject({
      status: 401,
      error: {
        code: 'AUTH_INVALID_CREDENTIALS',
        message: expect.stringContaining('Invalid credentials')
      }
    });
  });

  it('handles LDAP server unreachable with fallback error', async () => {
    // Arrange: Mock LDAP server unavailable
    server.use(
      http.post('http://localhost:7075/api/auth/login', () => {
        return HttpResponse.json<ErrorResponse>(
          {
            error: {
              code: 'AUTH_LDAP_UNAVAILABLE',
              message: 'LDAP server unreachable - authentication failed',
              correlationId: 'test-correlation-id',
              details: {
                ldapUrl: 'ldap://192.168.0.1',
                fallbackAttempted: true
              }
            }
          },
          { status: 503 }
        );
      })
    );

    // Act & Assert
    await expect(
      apiClient.login('dechawat', 'TestPassword123')
    ).rejects.toMatchObject({
      status: 503,
      error: {
        code: 'AUTH_LDAP_UNAVAILABLE'
      }
    });
  });
});

// =============================================================================
// POST /api/auth/login - SQL Authentication
// =============================================================================

describe('POST /api/auth/login - SQL', () => {
  it('returns LoginResponse schema on successful SQL authentication', async () => {
    // Arrange: Mock SQL authentication
    server.use(
      http.post('http://localhost:7075/api/auth/login', async ({ request }) => {
        const body = await request.json() as any;

        if (body.username === 'warehouse_user' && body.password === 'SqlPassword456') {
          return HttpResponse.json<LoginResponse>({
            token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.sql-auth-token',
            user: {
              userid: 50,
              username: 'warehouse_user',
              firstName: 'Warehouse',
              lastName: 'User',
              department: 'Operations',
              authSource: 'LOCAL',
              permissions: ['partial-picking']
            }
          });
        }

        return HttpResponse.json(
          { error: { code: 'AUTH_INVALID_CREDENTIALS', message: 'Invalid credentials', correlationId: 'test' } },
          { status: 401 }
        );
      })
    );

    // Act
    const response = await apiClient.login('warehouse_user', 'SqlPassword456');

    // Assert
    expect(response.user.authSource).toBe('LOCAL');
    expect(response.user.username).toBe('warehouse_user');
  });
});

// =============================================================================
// POST /api/auth/refresh - Token Refresh
// =============================================================================

describe('POST /api/auth/refresh', () => {
  it('returns new JWT token on successful refresh', async () => {
    // Arrange: Mock token refresh
    const oldToken = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.old-token';

    server.use(
      http.post('http://localhost:7075/api/auth/refresh', ({ request }) => {
        const authHeader = request.headers.get('Authorization');

        if (authHeader === `Bearer ${oldToken}`) {
          return HttpResponse.json({
            token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.new-token'
          });
        }

        return HttpResponse.json(
          { error: { code: 'AUTH_INVALID_TOKEN', message: 'Invalid token', correlationId: 'test' } },
          { status: 401 }
        );
      })
    );

    // Act
    const response = await apiClient.refreshToken(oldToken);

    // Assert
    expect(response.token).toBeDefined();
    expect(response.token).not.toBe(oldToken);
    expect(response.token).toMatch(/^eyJ/);
  });

  it('returns error on expired token', async () => {
    // Arrange: Mock expired token
    server.use(
      http.post('http://localhost:7075/api/auth/refresh', () => {
        return HttpResponse.json<ErrorResponse>(
          {
            error: {
              code: 'AUTH_INVALID_TOKEN',
              message: 'Token expired',
              correlationId: 'test-correlation-id',
              details: null
            }
          },
          { status: 401 }
        );
      })
    );

    // Act & Assert
    await expect(
      apiClient.refreshToken('expired-token')
    ).rejects.toMatchObject({
      status: 401,
      error: {
        code: 'AUTH_INVALID_TOKEN',
        message: expect.stringContaining('expired')
      }
    });
  });
});

// =============================================================================
// GET /api/auth/me - Current User
// =============================================================================

describe('GET /api/auth/me', () => {
  it('returns UserDTO schema for authenticated user', async () => {
    // Arrange: Mock authenticated request
    const token = 'valid-jwt-token';

    server.use(
      http.get('http://localhost:7075/api/auth/me', ({ request }) => {
        const authHeader = request.headers.get('Authorization');

        if (authHeader === `Bearer ${token}`) {
          return HttpResponse.json<UserDTO>({
            userid: 42,
            username: 'dechawat',
            firstName: 'Dechawat',
            lastName: 'Wongsirasawat',
            department: 'Warehouse',
            authSource: 'LDAP',
            permissions: ['partial-picking']
          });
        }

        return HttpResponse.json(
          { error: { code: 'AUTH_INVALID_TOKEN', message: 'Unauthorized', correlationId: 'test' } },
          { status: 401 }
        );
      })
    );

    // Act
    const user = await apiClient.getCurrentUser(token);

    // Assert
    expect(user.userid).toBe(42);
    expect(user.username).toBe('dechawat');
    expect(user.firstName).toBe('Dechawat');
    expect(user.authSource).toBe('LDAP');
  });

  it('returns error when no token provided', async () => {
    // Arrange: Mock unauthorized request
    server.use(
      http.get('http://localhost:7075/api/auth/me', () => {
        return HttpResponse.json<ErrorResponse>(
          {
            error: {
              code: 'AUTH_INVALID_TOKEN',
              message: 'No authorization token provided',
              correlationId: 'test-correlation-id',
              details: null
            }
          },
          { status: 401 }
        );
      })
    );

    // Act & Assert
    await expect(
      apiClient.getCurrentUser('')
    ).rejects.toMatchObject({
      status: 401,
      error: {
        code: 'AUTH_INVALID_TOKEN'
      }
    });
  });
});

File: backend/tests/contract/auth_contract_test.rs
// Authentication Contract Tests
// Tests validate implementation against contracts/openapi.yaml

use axum::http::StatusCode;
use serde_json::json;

mod test_helpers;
use test_helpers::*;

// =============================================================================
// POST /api/auth/login - LDAP Authentication
// =============================================================================

#[tokio::test]
async fn test_login_ldap_success() {
    // Arrange
    let app = create_test_app().await;
    let payload = json!({
        "username": "dechawat",
        "password": "TestPassword123"
    });

    // Act
    let response = app.post("/api/auth/login").json(&payload).send().await;

    // Assert: Response matches LoginResponse schema
    assert_eq!(response.status(), StatusCode::OK);

    let body: LoginResponse = response.json().await;
    assert!(!body.token.is_empty(), "JWT token should not be empty");
    assert_eq!(body.user.username, "dechawat");
    assert_eq!(body.user.auth_source, "LDAP");
    assert!(body
        .user
        .permissions
        .contains(&"partial-picking".to_string()));

    // Validate JWT token format
    assert!(body.token.starts_with("eyJ"), "JWT should start with eyJ");
}

#[tokio::test]
async fn test_login_ldap_invalid_credentials() {
    // Arrange
    let app = create_test_app().await;
    let payload = json!({
        "username": "dechawat",
        "password": "WrongPassword"
    });

    // Act
    let response = app.post("/api/auth/login").json(&payload).send().await;

    // Assert: Response matches ErrorResponse schema
    assert_eq!(response.status(), StatusCode::UNAUTHORIZED);

    let body: ErrorResponse = response.json().await;
    assert_eq!(body.error.code, "AUTH_INVALID_CREDENTIALS");
    assert!(body.error.message.contains("Invalid credentials"));
    assert!(!body.error.correlation_id.is_empty());
}

#[tokio::test]
async fn test_login_ldap_server_unreachable() {
    // Arrange: Mock LDAP server unreachable scenario
    let app = create_test_app_with_ldap_offline().await;
    let payload = json!({
        "username": "dechawat",
        "password": "TestPassword123"
    });

    // Act
    let response = app.post("/api/auth/login").json(&payload).send().await;

    // Assert: Should fallback to SQL authentication
    // This test validates FR-002 dual authentication with fallback
    assert_eq!(response.status(), StatusCode::UNAUTHORIZED);

    let body: ErrorResponse = response.json().await;
    assert_eq!(body.error.code, "AUTH_LDAP_UNAVAILABLE");
}

// =============================================================================
// POST /api/auth/login - SQL Authentication
// =============================================================================

#[tokio::test]
async fn test_login_sql_success() {
    // Arrange
    let app = create_test_app().await;
    let payload = json!({
        "username": "warehouse_user",
        "password": "SqlPassword456"
    });

    // Act
    let response = app.post("/api/auth/login").json(&payload).send().await;

    // Assert
    assert_eq!(response.status(), StatusCode::OK);

    let body: LoginResponse = response.json().await;
    assert_eq!(body.user.username, "warehouse_user");
    assert_eq!(body.user.auth_source, "LOCAL");

    // Validate bcrypt password hashing was used
    // (Implementation should compare against tbl_user.pword hash)
}

#[tokio::test]
async fn test_login_sql_invalid_password() {
    // Arrange
    let app = create_test_app().await;
    let payload = json!({
        "username": "warehouse_user",
        "password": "WrongPassword"
    });

    // Act
    let response = app.post("/api/auth/login").json(&payload).send().await;

    // Assert
    assert_eq!(response.status(), StatusCode::UNAUTHORIZED);

    let body: ErrorResponse = response.json().await;
    assert_eq!(body.error.code, "AUTH_INVALID_CREDENTIALS");
}

// =============================================================================
// POST /api/auth/refresh - Token Refresh
// =============================================================================

#[tokio::test]
async fn test_refresh_token_success() {
    // Arrange: Get valid JWT token first
    let app = create_test_app().await;
    let login_response = login_test_user(&app, "dechawat", "TestPassword123").await;
    let old_token = login_response.token;

    // Act
    let response = app
        .post("/api/auth/refresh")
        .header("Authorization", format!("Bearer {}", old_token))
        .send()
        .await;

    // Assert
    assert_eq!(response.status(), StatusCode::OK);

    let body: TokenResponse = response.json().await;
    assert!(!body.token.is_empty());
    assert_ne!(body.token, old_token, "New token should be different");
}

#[tokio::test]
async fn test_refresh_token_expired() {
    // Arrange: Use expired JWT token
    let app = create_test_app().await;
    let expired_token = create_expired_jwt_token();

    // Act
    let response = app
        .post("/api/auth/refresh")
        .header("Authorization", format!("Bearer {}", expired_token))
        .send()
        .await;

    // Assert
    assert_eq!(response.status(), StatusCode::UNAUTHORIZED);

    let body: ErrorResponse = response.json().await;
    assert_eq!(body.error.code, "AUTH_INVALID_TOKEN");
    assert!(body.error.message.contains("expired"));
}

// =============================================================================
// GET /api/auth/me - Current User Details
// =============================================================================

#[tokio::test]
async fn test_get_current_user_success() {
    // Arrange
    let app = create_test_app().await;
    let login_response = login_test_user(&app, "dechawat", "TestPassword123").await;

    // Act
    let response = app
        .get("/api/auth/me")
        .header("Authorization", format!("Bearer {}", login_response.token))
        .send()
        .await;

    // Assert
    assert_eq!(response.status(), StatusCode::OK);

    let body: UserDTO = response.json().await;
    assert_eq!(body.username, "dechawat");
    assert_eq!(body.first_name, Some("Dechawat".to_string()));
    assert_eq!(body.last_name, Some("Wongsirasawat".to_string()));
    assert_eq!(body.department, Some("Warehouse".to_string()));
}

#[tokio::test]
async fn test_get_current_user_no_token() {
    // Arrange
    let app = create_test_app().await;

    // Act
    let response = app.get("/api/auth/me").send().await;

    // Assert
    assert_eq!(response.status(), StatusCode::UNAUTHORIZED);

    let body: ErrorResponse = response.json().await;
    assert_eq!(body.error.code, "AUTH_INVALID_TOKEN");
}

// =============================================================================
// Type Definitions (matching OpenAPI schemas)
// =============================================================================

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
struct LoginResponse {
    token: String,
    user: UserDTO,
}

#[derive(Debug, Serialize, Deserialize)]
struct UserDTO {
    userid: i32,
    username: String,
    #[serde(rename = "firstName")]
    first_name: Option<String>,
    #[serde(rename = "lastName")]
    last_name: Option<String>,
    department: Option<String>,
    #[serde(rename = "authSource")]
    auth_source: String,
    permissions: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize)]
struct TokenResponse {
    token: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct ErrorResponse {
    error: ErrorDetails,
}

#[derive(Debug, Serialize, Deserialize)]
struct ErrorDetails {
    code: String,
    message: String,
    #[serde(rename = "correlationId")]
    correlation_id: String,
    details: Option<serde_json::Value>,
}
